File: backend/app/api/routes/llm.py
from fastapi import APIRouter, Depends, HTTPException
from sqlmodel import Session
from app.core.responses import ok
from app.db.session import get_session
from app.models.schemas import (
    IdeationRequest,
    AssetRequest,
    LandingPageRequest,
    ThankYouRequest,
    NurtureRequest,
    UpgradeOfferRequest,
    LinkedInRequest,
    HeroImageRequest,
    PersonaSummaryRequest,
    ChatRequest,
)
from app.services import settings as settings_service
from app.services.llm_service import (
    LLMClient,
    LLMProviderError,
    ideate_lead_magnets,
    generate_asset,
    generate_landing_page,
    generate_thank_you_page,
    generate_nurture_sequence,
    generate_upgrade_offer,
    generate_linkedin_post,
    generate_hero_image,
    generate_persona_summary,
    chat_marketing_assistant,
)

router = APIRouter()


def _client(session: Session) -> LLMClient:
    cfg = settings_service.get_app_settings(session)
    return LLMClient(cfg)


@router.post("/ideate", response_model=None)
async def ideate(payload: IdeationRequest, session: Session = Depends(get_session)):
    client = _client(session)
    try:
        ideas = await ideate_lead_magnets(
            client, 
            payload.icp, 
            product_context=payload.product_context,
            offer_type=payload.offer_type, 
            brand_voice=payload.brand_voice, 
            target_conversion=payload.target_conversion
        )
        return ok(ideas)
    except LLMProviderError as exc:
        raise HTTPException(
            status_code=exc.status_code,
            detail={"provider": exc.provider, "message": str(exc), "details": exc.details},
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/asset", response_model=None)
async def asset(payload: AssetRequest, session: Session = Depends(get_session)):
    client = _client(session)
    try:
        result = await generate_asset(
            client,
            payload.idea,
            payload.icp,
            product_context=payload.product_context,
            offer_type=payload.offer_type,
            brand_voice=payload.brand_voice,
        )
        return ok(result)
    except LLMProviderError as exc:
        raise HTTPException(
            status_code=exc.status_code,
            detail={"provider": exc.provider, "message": str(exc), "details": exc.details},
        )


@router.post("/landing-page", response_model=None)
async def landing_page(payload: LandingPageRequest, session: Session = Depends(get_session)):
    client = _client(session)
    try:
        result = await generate_landing_page(
            client, 
            payload.idea, 
            payload.asset, 
            payload.icp,
            product_context=payload.product_context,
            image_url=payload.image_url,
            offer_type=payload.offer_type,
            brand_voice=payload.brand_voice,
            target_conversion=payload.target_conversion
        )
        return ok(result)
    except LLMProviderError as exc:
        from fastapi import HTTPException
        raise HTTPException(
            status_code=exc.status_code,
            detail={"provider": exc.provider, "message": str(exc), "details": exc.details},
        )


@router.post("/thank-you", response_model=None)
async def thank_you(payload: ThankYouRequest, session: Session = Depends(get_session)):
    client = _client(session)
    try:
        result = await generate_thank_you_page(client, payload.idea)
        return ok(result)
    except LLMProviderError as exc:
        raise HTTPException(
            status_code=exc.status_code,
            detail={"provider": exc.provider, "message": str(exc), "details": exc.details},
        )


@router.post("/nurture-sequence", response_model=None)
async def nurture(payload: NurtureRequest, session: Session = Depends(get_session)):
    client = _client(session)
    try:
        result = await generate_nurture_sequence(
            client,
            payload.idea,
            asset=payload.asset,
            product_context=payload.product_context,
            strategy_summary=payload.strategy_summary,
            brand_voice=payload.brand_voice,
            target_conversion=payload.target_conversion,
        )
        return ok(result)
    except LLMProviderError as exc:
        raise HTTPException(
            status_code=exc.status_code,
            detail={"provider": exc.provider, "message": str(exc), "details": exc.details},
        )


@router.post("/upgrade-offer", response_model=None)
async def upgrade(payload: UpgradeOfferRequest, session: Session = Depends(get_session)):
    client = _client(session)
    try:
        result = await generate_upgrade_offer(
            client,
            payload.idea,
            payload.emails,
            product_context=payload.product_context,
            offer_type=payload.offer_type,
            brand_voice=payload.brand_voice,
            target_conversion=payload.target_conversion,
        )
        return ok(result)
    except LLMProviderError as exc:
        raise HTTPException(
            status_code=exc.status_code,
            detail={"provider": exc.provider, "message": str(exc), "details": exc.details},
        )


@router.post("/linkedin-post", response_model=None)
async def linkedin(payload: LinkedInRequest, session: Session = Depends(get_session)):
    client = _client(session)
    try:
        result = await generate_linkedin_post(
            client,
            payload.idea,
            payload.landing_page,
            brand_voice=payload.brand_voice,
            product_context=payload.product_context,
            strategy_summary=payload.strategy_summary,
        )
        return ok({"text": result})
    except LLMProviderError as exc:
        raise HTTPException(
            status_code=exc.status_code,
            detail={"provider": exc.provider, "message": str(exc), "details": exc.details},
        )


@router.post("/hero-image", response_model=None)
async def hero_image(payload: HeroImageRequest, session: Session = Depends(get_session)):
    client = _client(session)
    try:
        result = await generate_hero_image(
            client,
            payload.idea,
            payload.icp,
            brand_voice=payload.brand_voice,
            offer_type=payload.offer_type,
        )
        return ok({"url": result})
    except LLMProviderError as exc:
        raise HTTPException(
            status_code=exc.status_code,
            detail={"provider": exc.provider, "message": str(exc), "details": exc.details},
        )


@router.post("/persona-summary", response_model=None)
async def persona_summary(payload: PersonaSummaryRequest, session: Session = Depends(get_session)):
    client = _client(session)
    try:
        result = await generate_persona_summary(client, payload.icp)
        return ok(result)
    except LLMProviderError as exc:
        raise HTTPException(
            status_code=exc.status_code,
            detail={"provider": exc.provider, "message": str(exc), "details": exc.details},
        )


@router.post("/chat", response_model=None)
async def chat(payload: ChatRequest, session: Session = Depends(get_session)):
    from app.services.projects import get_project

    client = _client(session)
    try:
        project = get_project(session, payload.project_id) if payload.project_id else None
        reply = await chat_marketing_assistant(client, payload.message, project)
        return ok({"reply": reply})
    except LLMProviderError as exc:
        raise HTTPException(
            status_code=exc.status_code,
            detail={"provider": exc.provider, "message": str(exc), "details": exc.details},
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: frontend/services/llm.ts
import { post } from './api';
import { LeadMagnetIdea, GeneratedAsset, LandingPageConfig, ICPProfile, ProductContext } from '../types';

export const generateLeadMagnetIdeas = async (
    icp: ICPProfile,
    offerType?: string,
    brandVoice?: string,
    targetConversion?: string,
    productContext?: ProductContext
): Promise<LeadMagnetIdea[]> => {
    return post<LeadMagnetIdea[]>("/api/llm/ideate", { icp, product_context: productContext, offer_type: offerType, brand_voice: brandVoice, target_conversion: targetConversion });
};

export const generateAssetContent = async (
    idea: LeadMagnetIdea,
    icp: ICPProfile,
    offerType?: string,
    brandVoice?: string,
    productContext?: ProductContext
): Promise<GeneratedAsset> => {
    return post<GeneratedAsset>("/api/llm/asset", { idea, icp, offer_type: offerType, brand_voice: brandVoice, product_context: productContext });
};

export const generateLandingPage = async (
    idea: LeadMagnetIdea,
    asset: GeneratedAsset,
    icp: ICPProfile,
    imageUrl?: string,
    offerType?: string,
    brandVoice?: string,
    targetConversion?: string,
    productContext?: ProductContext
): Promise<LandingPageConfig> => {
    return post<LandingPageConfig>("/api/llm/landing-page", {
        idea,
        asset,
        icp,
        image_url: imageUrl,
        offer_type: offerType,
        brand_voice: brandVoice,
        target_conversion: targetConversion,
        product_context: productContext
    });
};

export const generateThankYouPage = async (idea: LeadMagnetIdea): Promise<any> => {
    return post<any>("/api/llm/thank-you", { idea });
};

export const generateNurtureSequence = async (
    idea: LeadMagnetIdea,
    brandVoice?: string,
    targetConversion?: string,
    productContext?: ProductContext,
    asset?: GeneratedAsset
): Promise<any> => {
    return post<any>("/api/llm/nurture-sequence", { idea, asset, brand_voice: brandVoice, target_conversion: targetConversion, product_context: productContext });
};

export const generateUpgradeOffer = async (
    idea: LeadMagnetIdea,
    emails: any[],
    offerType?: string,
    brandVoice?: string,
    productContext?: ProductContext,
    targetConversion?: string
): Promise<any> => {
    return post<any>("/api/llm/upgrade-offer", { idea, emails, offer_type: offerType, brand_voice: brandVoice, product_context: productContext, target_conversion: targetConversion });
};

export const generateLinkedInPost = async (
    idea: LeadMagnetIdea,
    landingPage: LandingPageConfig,
    brandVoice?: string,
    productContext?: ProductContext
): Promise<string> => {
    const res = await post<{text: string}>("/api/llm/linkedin-post", { idea, landing_page: landingPage, brand_voice: brandVoice, product_context: productContext });
     return res.text;
};

export const generateHeroImage = async (idea: LeadMagnetIdea, icp: ICPProfile, offerType?: string, brandVoice?: string): Promise<string> => {
     const res = await post<{url: string}>("/api/llm/hero-image", { idea, icp, offer_type: offerType, brand_voice: brandVoice });
     return res.url;
};

export const generatePersonaSummary = async (icp: ICPProfile): Promise<any> => {
    return post<any>("/api/llm/persona-summary", { icp });
};

export const analyzeWebsite = async (url: string): Promise<any> => {
    return post<any>("/api/generation/analyze-website", { url });
};
